File: ODMultiColumnLabel/ODMultiColumnLabel.m
//
//  ODMultiColumnLabel.m
//  ODMultiColumnLabel
//
//  Created by Fabio Ritrovato on 15/07/2014.
//  Copyright (c) 2014 orange in a day. All rights reserved.
//
// https://github.com/Sephiroth87/ODMultiColumnLabel
//

#import "ODMultiColumnLabel.h"

@interface ODMultiColumnLabel ()
{
    NSTextStorage *_textStorage;
    NSLayoutManager *_multicolumnManager;
    NSLayoutManager *_singlecolumnManager;
}

@end

@implementation ODMultiColumnLabel

- (instancetype)initWithFrame:(CGRect)frame {
    self = [super initWithFrame:frame];
    if (self) {
        [self commonInit];
    }
    return self;
}

- (instancetype)initWithCoder:(NSCoder *)aDecoder
{
    self = [super initWithCoder:aDecoder];
    if (self) {
        [self commonInit];
    }
    return self;
}

- (void)commonInit
{
    _textStorage = [NSTextStorage new];
    
    _multicolumnManager = [NSLayoutManager new];
    
    _singlecolumnManager = [NSLayoutManager new];
    NSTextContainer *singlecolumnContainer = [[NSTextContainer alloc] initWithSize:CGSizeZero];
    singlecolumnContainer.lineFragmentPadding = 0.0f;
    [_singlecolumnManager addTextContainer:singlecolumnContainer];
    
    [_textStorage addLayoutManager:_multicolumnManager];
    [_textStorage addLayoutManager:_singlecolumnManager];
    
    _numberOfColumns = 1;
    _columnsSpacing = 14.0f;
    [self updateText];
}

- (void)updateText
{
    if (!self.text || !_multicolumnManager) {
        return;
    }
    
    // Ensure text storage has at least all attributes from the label properties
    NSShadow *shadow = [NSShadow new];
    shadow.shadowOffset = self.shadowOffset;
    shadow.shadowColor = self.shadowColor;
    NSMutableParagraphStyle *paragraph = [NSMutableParagraphStyle new];
    paragraph.alignment = self.textAlignment;
    // BUG: with different line break modes at range 0, the text manager seems to layout only the first line fragment
    // and then stop, investigate more, for now, force WordWrapping
    // paragraph.lineBreakMode = self.lineBreakMode;
    paragraph.lineBreakMode = NSLineBreakByWordWrapping;
    NSDictionary *attributes = @{
                                 NSFontAttributeName: self.font,
                                 NSForegroundColorAttributeName: self.textColor,
                                 NSShadowAttributeName: shadow,
                                 NSParagraphStyleAttributeName: paragraph
                                 };
    NSMutableAttributedString *string = [[NSMutableAttributedString alloc] initWithString:self.text attributes:attributes];
    // Copy existing attributes, we do it after because they may override the label default values
    [self.attributedText enumerateAttributesInRange:NSMakeRange(0, [self.attributedText length]) options:0 usingBlock:^(NSDictionary *attrs, NSRange range, BOOL *stop) {
        [string addAttributes:attrs range:range];
        if (range.location == 0 && attrs[NSParagraphStyleAttributeName]) {
            NSMutableParagraphStyle *paragraph = [attrs[NSParagraphStyleAttributeName] mutableCopy];
            // Same as above
            paragraph.lineBreakMode = NSLineBreakByWordWrapping;
            [string addAttribute:NSParagraphStyleAttributeName value:paragraph range:range];
        }
    }];
    [_textStorage setAttributedString:string];
    
    CGFloat columnWidth = (self.frame.size.width - MAX(0, (NSInteger)_numberOfColumns - 1) * _columnsSpacing) / (float)_numberOfColumns;
    
    NSTextContainer *singlecolumnContainer = _singlecolumnManager.textContainers[0];
    singlecolumnContainer.size = CGSizeMake( columnWidth, CGFLOAT_MAX);
    [_singlecolumnManager glyphRangeForTextContainer:singlecolumnContainer];
    
    __block CGFloat columnHeight = 0.0f;
    CGFloat proposedColumnHeight = ceilf([_singlecolumnManager usedRectForTextContainer:singlecolumnContainer].size.height / (float)_numberOfColumns);
    [_singlecolumnManager enumerateLineFragmentsForGlyphRange:NSMakeRange(0, [_textStorage length]) usingBlock:^(CGRect rect, CGRect usedRect, NSTextContainer *textContainer, NSRange glyphRange, BOOL *stop) {
        columnHeight += rect.size.height;
        if (columnHeight >= proposedColumnHeight) {
            *stop = YES;
        }
    }];
    
    while ([_multicolumnManager.textContainers count] != _numberOfColumns) {
        if ([_multicolumnManager.textContainers count] < _numberOfColumns) {
            NSTextContainer *container = [[NSTextContainer alloc] initWithSize:CGSizeZero];
            container.lineFragmentPadding = 0;
            [_multicolumnManager addTextContainer:container];
            [_multicolumnManager glyphRangeForTextContainer:container];
        } else {
            [_multicolumnManager removeTextContainerAtIndex:[_multicolumnManager.textContainers count] - 1];
        }
    }
    
    [_multicolumnManager.textContainers enumerateObjectsUsingBlock:^(NSTextContainer *container, NSUInteger idx, BOOL *stop) {
        container.size = CGSizeMake(columnWidth, columnHeight);
    }];
    
    [self invalidateIntrinsicContentSize];
    [self setNeedsDisplay];
}

#pragma mark - Drawing

- (void)drawRect:(CGRect)rect
{
    CGFloat columnWidth = (self.frame.size.width - MAX(0, (NSInteger)_numberOfColumns - 1) * _columnsSpacing) / (float)_numberOfColumns;
    [_multicolumnManager.textContainers enumerateObjectsUsingBlock:^(NSTextContainer *container, NSUInteger idx, BOOL *stop) {
        CGPoint containerOrigin = CGPointMake(idx * (columnWidth + _columnsSpacing), 0.0f);
        NSRange containerRange = [_multicolumnManager glyphRangeForTextContainer:container];
        [_multicolumnManager drawBackgroundForGlyphRange:containerRange atPoint:containerOrigin];
        [_multicolumnManager drawGlyphsForGlyphRange:containerRange atPoint:containerOrigin];
    }];
}

#pragma mark - Layout

- (void)layoutSubviews
{
    [super layoutSubviews];
    [self updateText];
}

- (CGSize)intrinsicContentSize
{
    return [self sizeThatFits:CGSizeMake(self.frame.size.width, CGFLOAT_MAX)];
}

- (CGSize)sizeThatFits:(CGSize)size
{
    __block CGFloat height = 0.0f;
    [_multicolumnManager.textContainers enumerateObjectsUsingBlock:^(NSTextContainer *container, NSUInteger idx, BOOL *stop) {
        height = MAX(height, [_multicolumnManager usedRectForTextContainer:container].size.height);
    }];
    return CGSizeMake(size.width, height);
}

#pragma mark - Setters

- (void)setNumberOfColumns:(NSUInteger)numberOfColumns
{
    if (_numberOfColumns != numberOfColumns) {
        _numberOfColumns = numberOfColumns;
        [self updateText];
    }
}

- (void)setColumnsSpacing:(CGFloat)columnsSpacing
{
    _columnsSpacing = columnsSpacing;
    [self updateText];
}

- (void)setText:(NSString *)text
{
    [super setText:text];
    [self updateText];
}

- (void)setFont:(UIFont *)font
{
    [super setFont:font];
    [self updateText];
}

- (void)setTextColor:(UIColor *)textColor
{
    [super setTextColor:textColor];
    [self updateText];
}

- (void)setShadowColor:(UIColor *)shadowColor
{
    [super setShadowColor:shadowColor];
    [self updateText];
}

- (void)setShadowOffset:(CGSize)shadowOffset
{
    [super setShadowOffset:shadowOffset];
    [self updateText];
}

- (void)setTextAlignment:(NSTextAlignment)textAlignment
{
    [super setTextAlignment:textAlignment];
    [self updateText];
}

- (void)setLineBreakMode:(NSLineBreakMode)lineBreakMode
{
    [super setLineBreakMode:lineBreakMode];
    [self updateText];
}

- (void)setAttributedText:(NSAttributedString *)attributedText
{
    [super setAttributedText:attributedText];
    [self updateText];
}

@end

File: ODMultiColumnLabel/ODMultiColumnLabel.swift
//
//  ODMultiColumnLabel.swift
//  ODMultiColumnLabel
//
//  Created by Fabio Ritrovato on 31/07/2014.
//  Copyright (c) 2014 orange in a day. All rights reserved.
//
// https://github.com/Sephiroth87/ODMultiColumnLabel
//

import UIKit

class ODMultiColumnLabel: UILabel {
    
    var numberOfColumns: UInt = 1 { didSet { updateText() } }
    var columnsSpacing: CGFloat = 14.0 { didSet { updateText() } }
    
    private let textStorage: NSTextStorage
    private let multicolumnManager: NSLayoutManager
    private let singleColumnManager: NSLayoutManager
    
    override var text: String! { didSet { updateText() } }
    override var font: UIFont! { didSet { updateText() } }
    override var textColor: UIColor! { didSet { updateText() } }
    override var shadowColor: UIColor! { didSet { updateText() } }
    override var shadowOffset: CGSize { didSet { updateText() } }
    override var textAlignment: NSTextAlignment { didSet { updateText() } }
    override var lineBreakMode: NSLineBreakMode { didSet { updateText() } }
    override var attributedText: NSAttributedString! { didSet { updateText() } }

    override init(frame: CGRect) {
        textStorage = NSTextStorage()
        multicolumnManager = NSLayoutManager()
        singleColumnManager = NSLayoutManager()
        super.init(frame: frame)
        commonInit()
    }
    
    required init?(coder aDecoder: NSCoder) {
        textStorage = NSTextStorage()
        multicolumnManager = NSLayoutManager()
        singleColumnManager = NSLayoutManager()
        super.init(coder: aDecoder)
        commonInit()
    }
    
    private func commonInit() {
        let singlecolumnContainer = NSTextContainer(size: .zero)
        singlecolumnContainer.lineFragmentPadding = 0.0
        singleColumnManager.addTextContainer(singlecolumnContainer)
        
        textStorage.addLayoutManager(multicolumnManager)
        textStorage.addLayoutManager(singleColumnManager)
    }
    
    private func updateText() {
        if (text == nil) {
            return
        }
        
        var shadow = NSShadow()
        shadow.shadowOffset = shadowOffset;
        shadow.shadowColor = shadowColor;
        var paragraph = NSMutableParagraphStyle()
        paragraph.alignment = textAlignment
        // BUG: with different line break modes at range 0, the text manager seems to layout only the first line fragment
        // and then stop, investigate more, for now, force WordWrapping
        // paragraph.lineBreakMode = lineBreakMode
        paragraph.lineBreakMode = .byWordWrapping
        let attributes = [
            NSFontAttributeName: font,
            NSForegroundColorAttributeName: textColor,
            NSShadowAttributeName: shadow,
            NSParagraphStyleAttributeName: paragraph
        ] as [String : Any]
        var string = NSMutableAttributedString(string: text, attributes: attributes)
        // Copy existing attributes, we do it after because they may override the label default values
        attributedText.enumerateAttributes(in: NSRange(location: 0, length: attributedText.length), options: []) { (attrs, range, stop) in
            string.addAttributes(attrs, range: range)
            if range.location == 0 {
                if let paragraph = attrs[NSParagraphStyleAttributeName] as? NSParagraphStyle {
                    var newParagraph = paragraph.mutableCopy() as! NSMutableParagraphStyle
                    // Same as above
                    newParagraph.lineBreakMode = .byWordWrapping
                    string.addAttribute(NSParagraphStyleAttributeName, value: newParagraph, range: range)
                }
            }
        }
        textStorage.setAttributedString(string)
        
        let columnWidth = (frame.width - CGFloat(max(0, UInt(numberOfColumns - 1))) * columnsSpacing) / CGFloat(numberOfColumns)
        
        let singlecolumnContainer = singleColumnManager.textContainers[0] as NSTextContainer
        singlecolumnContainer.size = CGSize(width: columnWidth, height: CGFloat.greatestFiniteMagnitude)
        singleColumnManager.glyphRange(for: singlecolumnContainer)

        var columnHeight: CGFloat = 0.0
        let proposedColumnHeight = ceil(singleColumnManager.usedRect(for: singlecolumnContainer).height / CGFloat(numberOfColumns))
        singleColumnManager.enumerateLineFragments(forGlyphRange: NSRange(location: 0, length: textStorage.length)) { (rect, usedRect, textContainer, glyphRange, stop) in
            columnHeight += rect.height
            if columnHeight >= proposedColumnHeight {
                stop.initialize(to: true)
            }
        }

        while multicolumnManager.textContainers.count != Int(numberOfColumns) {
            if multicolumnManager.textContainers.count < Int(numberOfColumns) {
                let container = NSTextContainer(size: .zero)
                container.lineFragmentPadding = 0.0
                multicolumnManager.addTextContainer(container)
                multicolumnManager.glyphRange(for: container)
            } else {
                multicolumnManager.removeTextContainer(at: multicolumnManager.textContainers.count - 1)
            }
        }
     
        for container in multicolumnManager.textContainers as [NSTextContainer] {
            container.size = CGSize(width: CGFloat(columnWidth), height: columnHeight)
        }
        
        invalidateIntrinsicContentSize()
        setNeedsDisplay()
    }
    
    // MARK: - Drawing
    
    override func draw(_ rect: CGRect) {
        let columnWidth = (frame.width - CGFloat(max(0, UInt(numberOfColumns - 1))) * columnsSpacing) / CGFloat(numberOfColumns)
        for (index, container) in (multicolumnManager.textContainers as [NSTextContainer]).enumerated() {
            let containerOrigin = CGPoint(x: CGFloat(index) * (columnWidth + columnsSpacing), y: 0.0)
            let containerRange = multicolumnManager.glyphRange(for: container)
            multicolumnManager.drawBackground(forGlyphRange: containerRange, at: containerOrigin)
            multicolumnManager.drawGlyphs(forGlyphRange: containerRange, at: containerOrigin)
        }
    }
    
    // MARK: - Layout
    
    override func layoutSubviews() {
        super.layoutSubviews()
        updateText()
    }
    
    override var intrinsicContentSize: CGSize {
        return sizeThatFits(CGSize(width: frame.width, height: CGFloat.greatestFiniteMagnitude))
    }
    
    override func sizeThatFits(_ size: CGSize) -> CGSize {
        var height: CGFloat = 0.0
        for container in multicolumnManager.textContainers as [NSTextContainer] {
            height = max(height, multicolumnManager.usedRect(for: container).height)
        }
        return CGSize(width: size.width, height: height)
    }
    
}
